alerts/config.go

```diff
@@ -0,0 +1,158 @@
+
+	"github.com/tidepool-org/platform/data/blood/glucose"
+	"github.com/tidepool-org/platform/structure"
+)
+
+// Config models a user's desired alerts.
+type Config struct {
+	OwnerID         string         `json:"ownerID" bson:"ownerID"`
+	InvitorID       string         `json:"invitorID" bson:"invitorID"`
+	UrgentLow       *WithThreshold `json:"urgentLow,omitempty" bson:"urgentLow,omitempty"`
+	Low             *Deluxe        `json:"low,omitempty" bson:"low,omitempty"`
+	High            *Deluxe        `json:"high,omitempty" bson:"high,omitempty"`
+	NotLooping      *WithDelay     `json:"notLooping,omitempty" bson:"notLooping,omitempty"`
+	NoCommunication *WithDelay     `json:"noCommunication,omitempty" bson:"noCommunication,omitempty"`
+}
+
+func (c Config) Validate(validator structure.Validator) {
+	validator.String("ownerID", &c.OwnerID).LengthGreaterThan(0)
+	validator.String("invitorID", &c.InvitorID).LengthGreaterThan(0)
+	if c.Low != nil {
+		c.Low.Validate(validator)
+	}
+	if c.UrgentLow != nil {
+		c.UrgentLow.Validate(validator)
+	}
+	if c.High != nil {
+		c.High.Validate(validator)
+	}
+	if c.NotLooping != nil {
+		c.NotLooping.Validate(validator)
+	}
+	if c.NoCommunication != nil {
+		c.NoCommunication.Validate(validator)
+	}
+}
+
+// Base describes the minimum specifics of a desired alert.
+type Base struct {
+	// Enabled controls whether notifications should be sent for this alert.
+	Enabled bool
+	// Repeat is measured in minutes.
+	Repeat DurationMinutes `json:"repeat"`
+}
+
+func (b Base) Validate(validator structure.Validator) {
+	validator.Bool("enabled", &b.Enabled)
+	dur := b.Repeat.Duration()
+	validator.Duration("repeat", &dur).GreaterThan(0 * time.Minute)
+}
+
+// DelayMixin adds a configurable delay.
+type DelayMixin struct {
+	// Delay is measured in minutes.
+	Delay DurationMinutes `json:"delay,omitempty"`
+}
+
+func (d DelayMixin) Validate(validator structure.Validator) {
+	dur := d.Delay.Duration()
+	validator.Duration("delay", &dur).GreaterThan(0 * time.Minute)
+}
+
+// ThresholdMixin adds a configurable threshold.
+type ThresholdMixin struct {
+	// Threshold is compared the current value to determine if an alert should
+	// be triggered.
+	Threshold `json:"threshold"`
+}
+
+func (t ThresholdMixin) Validate(validator structure.Validator) {
+	t.Threshold.Validate(validator)
+}
+
+// WithThreshold extends Base with ThresholdMixin.
+type WithThreshold struct {
+	Base           `bson:",inline"`
+	ThresholdMixin `bson:",inline"`
+}
+
+func (d WithThreshold) Validate(validator structure.Validator) {
+	d.Base.Validate(validator)
+	d.ThresholdMixin.Validate(validator)
+}
+
+// WithDelay extends Base with DelayMixin.
+type WithDelay struct {
+	Base       `bson:",inline"`
+	DelayMixin `bson:",inline"`
+}
+
+func (d WithDelay) Validate(validator structure.Validator) {
+	d.Base.Validate(validator)
+	d.DelayMixin.Validate(validator)
+}
+
+// Deluxe extends Base with both DelayMixin and ThresholdMixin.
+type Deluxe struct {
```

Comment:
I have trouble understanding where the name of this struct comes from. Can you please clarify or revise?

Reply:
It's not a good name. Do you have a suggestion for something better?

Here "deluxe" is referring to the fact that this type of alert has all of the possible options, namely it has `DelayMixin` and `ThresholdMixin`.

Reply:
It's also just shorter than `BaseWithDelayAndThreshold`.